### ISSUE TITLE
Integer permute produces incorrect data

### ISSUE THREAD

**Issue by LPanosTT** [tenstorrent/tt-metal#19950]
The following test will fail:
```
def test_integer_permute(device):

    torch_input = torch.randint(-10, 10, (1, 2, 1, 1))
    torch_output = torch.permute(torch_input, [0, 2, 3, 1])

    input_tensor = ttnn.from_torch(torch_input, layout=ttnn.TILE_LAYOUT, dtype=ttnn.int32, device=device)
    output_tensor = ttnn.permute(input_tensor, [0, 2, 3, 1])
    output_tensor = ttnn.to_torch(output_tensor)

    assert torch_output.shape == output_tensor.shape
    assert torch.all(torch_output == output_tensor)
```

If permute isn't intended to be supported for integer data types yet, an informative error should be thrown stating so instead.

**Comment by jasondavies** [comment]
I can do this.  Can you explain how to approach this with NOC APIs as an alternative to RISC-V?

**Comment by Shubhamsaboo** [comment]
Hey @jasondavies!

Happy to assign you to this.

Be sure that you read over the [BOUNTY TERMS](https://docs.tenstorrent.com/bounty_terms.html), as you agree to all of them by picking up this issue.

If we haven't received any communication from you in a two week period, we will re-assign the issue, and you will no longer be eligible for completion of this bounty.

If you have any questions feel free to ping me here on this issue! Look forward to seeing your submission.

**Comment by jasondavies** [comment]
I'm happy to go the RISC approach but can you explain in a bit more detail why LLK transpose_wh doesn't work for int32?  It's a bit difficult to follow the code.  It feels like it would be simpler to fix transpose_wh if that's at all possible.

**Comment by jasondavies** [comment]
The issue with a baby RISC implementation is that the maximum write throughput is 32b per 5 cycles, which would mean 1280 cycles for each 16x16 face.  @corsix suggested splitting each face into 2x 16b 16x16 faces, transposing each face via SrcB transpose, then recombining to 32b.  This should be a lot faster than the baby RISC approach.

One approach could be to add new LLK APIs for splitting/recombining (they might be useful for other things).  Then it should be as simple as:

1. copy data to dest
2. split into two tiles
3. `_llk_math_transpose_dest_` on each tile
4. recombine two tiles
5. copy data from dest

**Comment by corsix** [comment]
> The issue with a baby RISC implementation is that the maximum write throughput is 32b per 5 cycles, which would mean 1280 cycles for each 16x16 face.

For reference, this is that baby RISCs on Wormhole, when writing directly to L1 using RISC `sw` instructions, have a maximum sustained write throughput of one 32b write every 5 cycles (whereas `sw` instructions to addresses _outside_ of L1 have a much more reasonable maximum sustained write throughput of one 32b write every cycle). One potential workaround for write-heavy workloads, when running as a compute kernel (rather than a data movement kernel) is to use the GPRs of the Tensix Scalar (ThCon) unit: an aligned 128b write to L1 can be done in five cycles; four cycles of `sw` instructions to write data to Tensix Scalar GPRs, followed by one cycle to issue a Tensix Scalar 128b write instruction (which'll move from Tensix Scalar GPRs to L1). A bunch of really subtle synchronisation is required to make this work though (there are some non-subtle ways of doing the synchronisation, but they'll tank performance), so I'd only recommend exploring this approach if the two 16b planes approach doesn't work.

**Comment by corsix** [comment]
And if the SrcB transpose route doesn't work, I think another viable route is:
1. Use unpackers to copy a tile of 32b data from L1 to Dst.
2. Use some sneaky _packer_ configuration to perform the transpose whilst copying the data from Dst back to L1.

I think step 2 in the above would involve 64 single-cycle pack instructions per 16x16 face, which is a lot faster than the 1280 cycles needed to do the operation purely on the baby RISCs, but probably still slower than the SrcB transpose route.

**Comment by jasondavies** [comment]
I've added support for 32-bit types to the transpose_dest LLK: tenstorrent/tt-llk#135.

**Comment by jasondavies** [comment]
Currently stuck on trying to get a 32-bit tile copied into Dst prior to transpose_dest, within this [transpose_wh](https://github.com/tenstorrent/tt-metal/blob/8a48d92c2b40e0bf8d6988fc3aff730409d634e4/ttnn/cpp/ttnn/operations/data_movement/transpose/device/kernels/compute/transpose_wh.cpp) ttnn kernel.

Some combination of llk_unpack_A and llk_math_eltwise_unary_datacopy?

**Comment by jasondavies** [comment]
Just confirming that my proof-of-concept in #21270 works.  The tricky part now is how to actually swap a 32-bit implementation into transpose_wh for 32-bit types.  Can anyone advise?

Roughly speaking, I think I want to get the data type/format for the CB index within transpose_wh, and use `copy_tile` + `transpose_wh_dest` for 32-bit data types.

Is it as simple as using `unpack_src_format[operand_id]`?  

This would be the simplest approach as `transpose_wh` is used in many transpose-related places (RM, sharded, etc.) and changing all of them to add a 32-bit case sounds like a lot of work.

**Comment by jasondavies** [comment]
I tried the above and it seems to work fine for tiled layouts.  Unfortunately not so for RM layouts -- unclear why at this point.  Maybe there's a hardcoded 16-bit size somewhere in the code?

**Comment by jasondavies** [comment]
Both PRs are ready for review/feedback:

- tenstorrent/tt-llk#135 - this needs to be reviewed and merged to support generic 32-bit transpose on Dst.
- #21270 - this uses the above 32-bit support to allow `transpose_wh_tile` to use `copy_tile` followed by a 32-bit transpose on Dst for the int32 case.  Note: row major layouts do not currently work correctly - either this needs to be diagnosed and fixed, or an error should be thrown.

**Comment by jasondavies** [comment]
Bit of a wild guess, but the int32 issues with row major layouts could be related to int32 issues with tilize/untilize (which are utilised by [transpose_xw_rm_single_tile](https://github.com/tenstorrent/tt-metal/blob/2b9aa799d371c4a43b7b1c850b4fad52b77a0cc9/ttnn/cpp/ttnn/operations/data_movement/permute/device/kernels/compute/transpose_xw_rm_single_tile_size.cpp); see #16860.

**Comment by jasondavies** [comment]
Would be great to get some help with figuring out why #21270 doesn't fully work for RM layouts.

**Comment by jasondavies** [comment]
> So, to support RM layout, you would need to first tilize it, transpose, and then untilize it.

Yes, I was expecting my modifications to transpose_wh_tile to work in conjunction with kernels like [transpose_xw_rm_single_tile_size](https://github.com/tenstorrent/tt-metal/blob/892bd593579f17488e20489bd2808ac53299e1b1/ttnn/cpp/ttnn/operations/data_movement/permute/device/kernels/compute/transpose_xw_rm_single_tile_size.cpp) (you can see there's tilize_block and pack_untilize_dst used there).  I tried several things last week but maybe I don't completely understand what's going wrong in there.

**Comment by jasondavies** [comment]
@nvelickovicTT It's this simple test.  If you change layout to `ttnn.TILE_LAYOUT`, it will pass (no assertion error).

```python
import ttnn
import torch

torch.manual_seed(0)
device = ttnn.open_device(device_id=0)

def test_integer_permute(device):
    permutation = [0, 2, 1]
    #shape = (2, 2, 2)
    shape = (64, 64, 64)
    #torch_input = torch.randint(-2**31, 2**31, shape)
    #torch_input = torch.full(shape, 0x7fff0001)
    #torch_input = torch.full(shape, 1)
    torch_input = torch.randint(-10, 10, shape)
    torch_output = torch.permute(torch_input, permutation)

    input_tensor = ttnn.from_torch(torch_input, layout=ttnn.ROW_MAJOR_LAYOUT, dtype=ttnn.int32, device=device)
    output_tensor = ttnn.permute(input_tensor, permutation)
    output_tensor = ttnn.to_torch(output_tensor)

    print("shapes", torch_output.shape, output_tensor.shape)
    print("tensors", torch_output, output_tensor)

    assert torch_output.shape == output_tensor.shape
    assert torch.all(torch_output == output_tensor)

try:
    test_integer_permute(device)
finally:
    ttnn.close_device(device)
```

**Comment by jasondavies** [comment]
Some elements get transposed, and some don't.  You can kind of see this if you use `torch_input = torch.full(shape, 0)`.  You will see some values are the old randint(-10, 10), and others correctly get written with zero.

It's almost like something's not working correctly in tilize or untilize.  I think I also checked just a basic copy_tile instead of transpose_wh_tile to see if that worked OK in that kernel.  I can't remember if that was successful.  I spent a lot of time trying to figure out if some parameters were 16-bit specific (e.g. calculating CB sizes etc.) further up the call stack but couldn't find anything there.

**Comment by nvelickovicTT** [comment]
OK, will take a look at the test tomorrow. Thanks for sharing.

**Comment by nvelickovicTT** [comment]
@jasondavies I looked at the test, and there doesn't seem to be anything LLK related that looks suspicious.
What I noticed is that in `ttnn/cpp/ttnn/operations/data_movement/permute/device/kernels/compute/transpose_xw_rm_single_tile_size.cpp`, when `num_blocks > 1`, that's when result starts to be wrong for int32.
However, when I try with float16_b I always get incorrect result.
More so, there is not a convenient way to force the test to run on single core, which is needed to efficiently debug a compute issue.

@LPanosTT , @sjameelTT can you help with modifying this test to run on single core?

**Comment by sjameelTT** [comment]
`sjameel/transpose_single_core` this will make all row major permute variants single core.

I noticed when switching to single core, that some of the other tests in the suite have PCC drops.

`tests/ttnn/unit_tests/operations/test_permute.py::test_permute_5d_blocked` has several. It seems like with enough iterations of the LLK PCC ends up dropping.

**Comment by jasondavies** [comment]
> `sjameel/transpose_single_core` this will make all row major permute variants single core.

Thanks for this.  I'm still stumped by this problem -- even removing transpose_wh_tile entirely and replacing with the following seems to work fine for bfloat16 and float32, but for int32 _seems_ to generate garbage for odd tiles only:

```diff
--- a/ttnn/cpp/ttnn/operations/data_movement/permute/device/kernels/compute/transpose_xw_rm_single_tile_size.cpp
+++ b/ttnn/cpp/ttnn/operations/data_movement/permute/device/kernels/compute/transpose_xw_rm_single_tile_size.cpp
@@ -39,11 +39,11 @@ void MAIN {

         // transpose input
         cb_wait_front(cb_tilize, 1);
-        transpose_wh_init_short(cb_tilize);
+        copy_tile_to_dst_init_short(cb_tilize);
         pack_untilize_dst_init_short<1>(cb_out);

         tile_regs_acquire();
-        transpose_wh_tile(cb_tilize, 0, 0);  // transpose call
+        copy_tile(cb_tilize, 0, 0);
         tile_regs_commit();

         // pack and untilize
```

Maybe kind of address/offset issue at the LLK level with unpack to dest?

**Comment by sjameelTT** [comment]
Definitely sounds like some sort of config problem to me. @nvelickovicTT have you made any progress debugging?

**Comment by jasondavies** [comment]
@nvelickovicTT FYI this is my latest test for the above copy_tile modification (so it doesn't actually do a transpose).

It does one test with bfloat16, and then another with int32, printing the non-matching elements and the first two dimensions.  This shows that everything in `ttnn_output[1]` and `ttnn_output[3]` doesn't match the expected output for int32.

```python
import ttnn
import torch

torch.manual_seed(0)
device = ttnn.open_device(device_id=0)

def test_integer_permute(device, is_32bit):
    permutation = [0, 2, 1]
    if is_32bit:
        dtype = ttnn.int32
        shape = (5, 32, 32)
        torch_input = torch.randint(0, 10, shape)
    else:
        dtype = ttnn.bfloat16
        shape = (2 * 5, 32, 32)
        torch_input = torch.randint(-10, 0, shape).bfloat16()

    torch_output = torch_input # torch.permute(torch_input, permutation)

    input_tensor = ttnn.from_torch(torch_input, layout=ttnn.ROW_MAJOR_LAYOUT, dtype=dtype, device=device)
    ttnn_output = ttnn.permute(input_tensor, permutation)
    ttnn_output = ttnn.to_torch(ttnn_output)

    if not is_32bit:
        ttnn_output = ttnn_output.bfloat16()

    for i, a in enumerate(torch_output):
        for j, row in enumerate(a):
            if not torch.all(ttnn_output[i][j] == row):
                print(i, j, ttnn_output[i][j], row)

    assert torch_output.shape == ttnn_output.shape
    assert torch.all(torch_output == ttnn_output)

try:
    test_integer_permute(device, 0)
    test_integer_permute(device, 1)
finally:
    ttnn.close_device(device)
```

**Comment by jasondavies** [comment]
I've narrowed the issue down to `pack_untilize` by replacing it with a plain `pack`, which doesn't seem to mangle the data:

- remove `pack_untilize_dst_init_short<1>(cb_out);`
- replace `pack_untilize_dst<1>(cb_out);` with `pack_tile(0, cb_out);`
- remove `pack_untilize_uninit(cb_out);`

**Comment by Shubhamsaboo** [comment]
Since this is turning out to be more difficult than what we expected so we are increasing the reward amount for this one to $3000. All the best Jason!

**Comment by jasondavies** [comment]
I tested and your fix works with my transpose_wh_tile permute RM test.  Thank you so much!

Will double-check, consolidate any other changes, and update shortly.

Does there need to be a similar change made for Blackhole?

**Comment by jasondavies** [comment]
@nvelickovicTT Just rebased my tt-metal PR #21270 against latest main.  This contains the important transpose_wh_tile change that checks for 32-bit inputs:

```cpp
ALWI void transpose_wh_tile(uint32_t icb, uint32_t itile, uint32_t idst) {
#if defined(TRISC_MATH) || defined(TRISC_UNPACK)
    const std::uint32_t src_format = get_operand_src_format(icb);
    const bool is_32bit = (src_format & 0xf) == (std::uint32_t)DataFormat::Int32;
    if (is_32bit) {
        copy_tile_init(icb);
        copy_tile(icb, itile, idst);

        transpose_wh_dest_init_short<true>();
        transpose_wh_dest<true>(idst);
    } else {
        UNPACK((llk_unpack_A<BroadcastType::NONE, false>(icb, itile, false)));
        MATH((llk_math_eltwise_unary_datacopy<A2D, BroadcastType::NONE, DST_ACCUM_MODE>(idst)));
    }
#endif
}
```

Though it does feel a bit like this could be simplified e.g. reference specific LLKs rather than calling copy_tile_init and copy_tile, it does solve the issue for all users of transpose_wh_tile.

Suggestions welcome for improvements on this front.

**Comment by nvelickovicTT** [comment]
@jasondavies that's great news. Thanks for pursuing this debug.
Blackhole implementation uses 2 contexts, so no need for a fix like this. Wormhole implementation had some issues with the 2nd context for unpack to dest tilize for int32, so that is why it diverges from the standard approach of using 2 config contexts.
We will revisit this for Wormhole at some point, but it is deprioritized at the moment.

That being said, it's obvious LLK code architecture doesn't gracefully handle mixed usage of 1 and 2 config contexts. It's something to think about on our side.

**Comment by jasondavies** [comment]
An update for anyone not aware of the various other discussions in other PRs:

- #22298 There is a pre-existing problem (possible race condition) with permute with bfloat16 on large enough tensors/shapes.  See the issue for more information: it's 100% reproducible when restricted to a single core, and adding a few hundred NOPs to the kernel makes the tests pass.

The same issue affects int32 on similar sizes/shapes of input tensors.

I'd recommend prioritising a fix for this as it affects existing bfloat16 uses of permute, as well as int32.

This can be added as a sub-issue @sjameelTT

**Comment by jasondavies** [comment]
I opened yet another issue here, this time it looks like I have a fix (which hopefully can be more fully explained/improved):

- #22383

**Comment by Shubhamsaboo** [comment]
Congrats Jaon on completing another bounty. Keep an eye on your inbox while we reach out for the payment.
